### Financial Performance and Profitability Analysis

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('datasets/bss_sales_data.csv')
df.shape

(140690, 20)

In [3]:
df.isnull().sum()

sku                           433
salesdate                     433
price                         433
unitsordered                  433
sales                         433
cogs                          433
fba                           433
reffee                        433
adspend                       433
profit                        433
comp_1_price                50299
comp_2_price                85411
comp_3_price               116972
comp_4_price               132411
comp_5_price               138654
comp_data_min_price         43121
comp_data_max_price         43121
managed_fba_stock_level     24740
min_price                     433
max_price                     433
dtype: int64

In [4]:
df = df.dropna(how='all').copy()

In [5]:
df.shape

(140257, 20)

In [6]:
df.isnull().sum().sort_values(ascending=False)

comp_5_price               138221
comp_4_price               131978
comp_3_price               116539
comp_2_price                84978
comp_1_price                49866
comp_data_min_price         42688
comp_data_max_price         42688
managed_fba_stock_level     24307
unitsordered                    0
price                           0
salesdate                       0
sku                             0
adspend                         0
profit                          0
fba                             0
reffee                          0
cogs                            0
sales                           0
min_price                       0
max_price                       0
dtype: int64

In [7]:
df.duplicated().sum()

np.int64(0)

In [8]:
df[['sku', 'salesdate']].duplicated().sum()

np.int64(596)

##### Checking if 596 sku records are duplicates or separate records

In [9]:
dup = df[df[['sku', 'salesdate']].duplicated(keep=False)].sort_values(['sku', 'salesdate'])
dup.head(20)

,sku,salesdate,price,unitsordered,sales,cogs,fba,reffee,adspend,profit,comp_1_price,comp_2_price,comp_3_price,comp_4_price,comp_5_price,comp_data_min_price,comp_data_max_price,managed_fba_stock_level,min_price,max_price
9303,Classification Folders SKU 11,3/19/2022,22.79,9.0,205.0,68.0,59.0,31.0,13.0,34.0,20.51,NaN,NaN,NaN,NaN,20.51,20.51,NaN,20.99,28.99
9304,Classification Folders SKU 11,3/19/2022,21.89,43.0,941.0,324.0,284.0,141.0,13.0,179.0,20.51,NaN,NaN,NaN,NaN,20.51,20.51,NaN,20.99,28.99
59905,File Folders SKU 39,1/1/2022,29.99,1.0,30.0,14.0,8.0,4.0,0.0,4.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,29.99,39.99
60493,File Folders SKU 39,1/1/2022,24.69,4.0,99.0,39.0,33.0,15.0,0.0,12.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,24.99,35.99
60235,File Folders SKU 39,1/1/2023,0.00,0.0,0.0,0.0,0.0,0.0,1.0,-1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,493.0,29.99,39.99
60850,File Folders SKU 39,1/1/2023,0.00,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,251.0,24.99,35.99
59913,File Folders SKU 39,1/10/2022,29.79,4.0,119.0,58.0,32.0,18.0,0.0,11.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,29.99,39.99
60501,File Folders SKU 39,1/10/2022,22.79,12.0,273.0,116.0,99.0,41.0,0.0,17.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,24.99,35.99
60243,File Folders SKU 39,1/10/2023,29.79,4.0,119.0,58.0,32.0,18.0,5.0,6.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,468.0,29.99,39.99
60858,File Folders SKU 39,1/10/2023,28.79,12.0,345.0,116.0,99.0,52.0,4.0,74.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,169.0,24.99,35.99


In [10]:
dup.groupby(['sku', 'salesdate']).size().value_counts()

2    596
Name: count, dtype: int64

In [11]:
dup.groupby(['sku', 'salesdate']).agg(
    rows=('sku', 'size'),
    total_units=('unitsordered', 'sum'),
    total_sales=('sales', 'sum'),
    total_profit=('profit', 'sum')
).sort_values('rows', ascending=False).head(20)

rows  total_units  total_sales  \
sku                           salesdate                                   
Classification Folders SKU 11 3/19/2022     2         52.0       1146.0   
File Folders SKU 39           1/1/2022      2          5.0        129.0   
                              1/1/2023      2          0.0          0.0   
                              1/10/2022     2         16.0        392.0   
                              1/10/2023     2         16.0        464.0   
                              1/11/2022     2         22.0        552.0   
                              1/11/2023     2         18.0        520.0   
                              1/12/2022     2         17.0        463.0   
                              1/12/2023     2          9.0        263.0   
                              1/13/2022     2          8.0        214.0   
                              1/13/2023     2         14.0        377.0   
                              1/14/2023     2          2.0         59.0   
                              1/15/2023     2          0.0          0.0   
                              1/16/2023     2         15.0        439.0   
                              1/17/2022     2          5.0        128.0   
                              1/17/2023     2         18.0        522.0   
                              1/18/2022     2         22.0        570.0   
                              1/18/2023     2         15.0        437.0   
                              1/19/2022     2         16.0        447.0   
                              1/19/2023     2          9.0        262.0   

                                         total_profit  
sku                           salesdate                
Classification Folders SKU 11 3/19/2022         213.0  
File Folders SKU 39           1/1/2022           16.0  
                              1/1/2023           -1.0  
                              1/10/2022          28.0  
                              1/10/2023          80.0  
                              1/11/2022          67.0  
                              1/11/2023          98.0  
                              1/12/2022          53.0  
                              1/12/2023          28.0  
                              1/13/2022          26.0  
                              1/13/2023          47.0  
                              1/14/2023           9.0  
                              1/15/2023           0.0  
                              1/16/2023          58.0  
                              1/17/2022          15.0  
                              1/17/2023         102.0  
                              1/18/2022          64.0  
                              1/18/2023          70.0  
                              1/19/2022          49.0  
                              1/19/2023          30.0

In [12]:
df.dtypes

sku                         object
salesdate                   object
price                      float64
unitsordered               float64
sales                      float64
cogs                       float64
fba                        float64
reffee                     float64
adspend                    float64
profit                     float64
comp_1_price               float64
comp_2_price               float64
comp_3_price               float64
comp_4_price               float64
comp_5_price               float64
comp_data_min_price        float64
comp_data_max_price        float64
managed_fba_stock_level    float64
min_price                  float64
max_price                  float64
dtype: object

In [13]:
df[['salesdate']].head()

,salesdate
0,1/1/2022
1,1/2/2022
2,1/3/2022
3,1/4/2022
4,1/5/2022


In [14]:
df[['price', 'unitsordered', 'sales', 'cogs', 'fba','reffee', 'adspend', 'profit']].describe()

,price,unitsordered,sales,cogs,fba,reffee,adspend,profit
count,140257.000000,140257.000000,140257.000000,140257.000000,140257.000000,140257.000000,140257.000000,140257.000000
mean,22.235573,19.373265,488.155258,160.097058,140.463150,73.220117,30.906814,83.468119
std,13.192553,31.634210,813.708806,285.708228,227.405391,122.055192,64.814287,192.928220
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,-4379.000000
25%,15.890000,2.000000,45.000000,15.000000,15.000000,7.000000,0.000000,0.000000
50%,24.590000,8.000000,182.000000,61.000000,53.000000,27.000000,7.000000,18.000000
75%,28.490000,23.000000,576.000000,186.000000,167.000000,86.000000,30.000000,86.000000
max,111.490000,833.000000,20900.000000,9654.000000,6847.000000,3135.000000,1628.000000,6753.000000


##### Convert Data Types

###### For salesdate

In [15]:
df['salesdate'] = pd.to_datetime(df['salesdate'])

In [16]:
df['salesdate'].dtype

dtype('<M8[ns]')

In [17]:
print("Start:", df['salesdate'].min())
print("End:", df['salesdate'].max())

Start: 2022-01-01 00:00:00
End: 2023-09-23 00:00:00


###### For unitsordered

In [19]:
#checking whether it contains decimals
df[df['unitsordered'] % 1 != 0]['unitsordered'].count()

np.int64(0)

In [20]:
df['unitsordered'] = df['unitsordered'].astype(int)

In [21]:
df['unitsordered'].dtype

dtype('int64')

In [22]:
calculated_profit = (
    df['sales']
    - df['cogs']
    - df['fba']
    - df['reffee']
    - df['adspend']
)

(df['profit'] - calculated_profit).abs().describe()

count    140257.0
mean          0.0
std           0.0
min           0.0
25%           0.0
50%           0.0
75%           0.0
max           0.0
dtype: float64

In [23]:
(df['profit'] != calculated_profit).sum()

np.int64(0)

##### Checking Zero sales records

In [24]:
df[(df['price'] == 0) | (df['sales'] == 0)].shape

(20220, 20)

In [25]:
df[df['price'] == 0][['sku', 'salesdate', 'price', 'unitsordered', 'sales','cogs', 'fba', 'reffee', 'adspend', 'profit']].head(20)

,sku,salesdate,price,unitsordered,sales,cogs,fba,reffee,adspend,profit
676,Binder Dividers SKU 10,2022-03-12,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
679,Binder Dividers SKU 10,2022-03-15,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
682,Binder Dividers SKU 10,2022-03-18,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
683,Binder Dividers SKU 10,2022-03-19,0.0,0,0.0,0.0,0.0,0.0,0.0,0.0
685,Binder Dividers SKU 10,2022-03-21,0.0,0,0.0,0.0,0.0,0.0,0.0,0.0
691,Binder Dividers SKU 10,2022-03-27,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
693,Binder Dividers SKU 10,2022-03-29,0.0,0,0.0,0.0,0.0,0.0,3.0,-3.0
696,Binder Dividers SKU 10,2022-04-01,0.0,0,0.0,0.0,0.0,0.0,2.0,-2.0
697,Binder Dividers SKU 10,2022-04-02,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
698,Binder Dividers SKU 10,2022-04-03,0.0,0,0.0,0.0,0.0,0.0,0.0,0.0


In [26]:
df[df['sales'] == 0][['sku', 'salesdate', 'price', 'unitsordered', 'sales','cogs', 'fba', 'reffee', 'adspend', 'profit']].head(20)

,sku,salesdate,price,unitsordered,sales,cogs,fba,reffee,adspend,profit
676,Binder Dividers SKU 10,2022-03-12,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
679,Binder Dividers SKU 10,2022-03-15,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
682,Binder Dividers SKU 10,2022-03-18,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
683,Binder Dividers SKU 10,2022-03-19,0.0,0,0.0,0.0,0.0,0.0,0.0,0.0
685,Binder Dividers SKU 10,2022-03-21,0.0,0,0.0,0.0,0.0,0.0,0.0,0.0
691,Binder Dividers SKU 10,2022-03-27,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
693,Binder Dividers SKU 10,2022-03-29,0.0,0,0.0,0.0,0.0,0.0,3.0,-3.0
696,Binder Dividers SKU 10,2022-04-01,0.0,0,0.0,0.0,0.0,0.0,2.0,-2.0
697,Binder Dividers SKU 10,2022-04-02,0.0,0,0.0,0.0,0.0,0.0,1.0,-1.0
698,Binder Dividers SKU 10,2022-04-03,0.0,0,0.0,0.0,0.0,0.0,0.0,0.0


###### Checking Negative Values